parsers/FO.py

```diff
@@ -22,97 +22,72 @@
-
-
 def fetch_production(
-    zone_key: VALID_ZONE_KEYS = "FO",
+    zone_key: ZoneKey = "FO",
     session: Session | None = None,
     target_datetime: datetime | None = None,
-    logger: Logger = getLogger("FO"),
-) -> dict:
-    if target_datetime:
+    logger: Logger = getLogger(__name__),
+) -> list[dict]:
+    if zone_key not in VALID_ZONE_KEYS:
+        raise ParserException(
+            PARSER,
+            f"This parser cannot retrieve data for zone {zone_key!r}",
+        )
+
+    if target_datetime is not None:
         # There is a API endpoint at https://www.sev.fo/api/elproduction/last7days
         # but it's currently returning nothing at all. Last checked on 2022-08-09
-        raise NotImplementedError("This parser is not yet able to parse past dates")
+        raise ParserException(
+            PARSER, "This parser is not yet able to parse past dates", zone_key=zone_key
+        )
 
     ses = session or Session()
     url = "https://www.sev.fo/api/realtimemap/now"
     response: Response = ses.get(url)
     obj = response.json()
 
-    data = {
-        "zoneKey": zone_key,
-        "capacity": {},
-        "production": {
-            "biomass": 0,
-            "coal": 0,
-            "gas": 0,
-            "geothermal": 0,
-            "nuclear": 0,
-            "solar": 0,
-            "unknown": 0,
-        },
-        "storage": {},
-        "source": "sev.fo",
-    }
+    production_breakdown_list = ProductionBreakdownList(logger)
+    production_mix = ProductionMix()
     for key, value in obj.items():
-        if key == "tiden":
-            data["datetime"] = datetime.fromisoformat(value).replace(tzinfo=FO)
         if "Sum" in key or "Test" in key or "VnVand" in key:
             # "VnVand" is the sum of hydro (Mýrarnar + Fossá + Heygar)
             continue
-        elif key.endswith(ZONE_MAP[zone_key]["data_key"]):
+
+        elif key.endswith(ZONE_KEY_TO_DATA_KEY[zone_key]):
             # E stands for Energy
-            raw_generation_type: str = key.replace(ZONE_MAP[zone_key]["data_key"], "")
-            generation_type = map_generation_type(raw_generation_type)
-            if not generation_type:
+            raw_generation_type: str = key.replace(ZONE_KEY_TO_DATA_KEY[zone_key], "")
+            generation_type = MAP_GENERATION.get(raw_generation_type)
+            if generation_type is None:
                 raise ParserException(
-                    "FO.py", f"Unknown generation type: {raw_generation_type}", zone_key
+                    PARSER, f"Unknown generation type: {raw_generation_type}", zone_key
                 )
             # Power (MW)
             value = float(value.replace(",", "."))
-            data["production"][generation_type] = (
-                data["production"].get(generation_type, 0) + value
-            )
+            production_mix.add_value(generation_type, value)
         else:
-            # print 'Unhandled key %s' % key
             pass
```

Comment:
Pretty sure we can just delete these two lines. They do nothing.

Reply:
In addition to production values aggregated by production mode that we handle above there are some extra keys which make their way through here (I think they are the itemised production values by power plant):

```
tiden
SdDiesel_E
FoVand_E
MyVand_E
HeVand_E
EiVand_E
StDiesel_E
StVand_E
NoDiesel_E
SorDiesel_E
HhVind_E
NeVind_E
VgDiesel_E
BoVand_E
SuSol_E
PoVind_E
RoVind_E
ViRoVind_E
SuStVind_E
NoStVind_E
SuVind_E
Minesto_E
Forka_E
VandH_E
OlieH_E
VindH_E
BiogasH_E
TidalH_E
VandSev_E
OlieSev_E
VindSev_E
SolSev_E
BiogasSev_E
TidalSev_E
SdDiesel_P
FoVand_P
MyVand_P
HeVand_P
EiVand_P
StDiesel_P
StVand_P
NoDiesel_P
SorDiesel_P
HhVind_P
NeVind_P
RoVind_P
ViRoVind_P
SuStVind_P
NoStVind_P
SuVind_P
VgDiesel_P
BoVand_P
SuSol_P
PoVind_P
Minesto_P
Forka_P
VandH_P
VindH_P
OlieH_P
BiogasH_P
TidalH_P
VandS_P
OlieS_P
VindS_P
SolS_P
VandSev_P
OlieSev_P
VindSev_P
SolSev_P
BiogasSev_P
TidalSev_P
Gront_P
Vatnid
Lomundaroyri
Heygadalur
Eidisvatn
Strandadalur
Ryskisvatn
Vatnsnesvatn
```

Reply:
Ah okey but then it should probably be continue instead of pass IMO as pass is usually a placeholder.